### PR TITLE
Mindy/extensions

### LINKED ISSUE #419: Refactor View Controllers
**Description:**
Break out protocol implementations into extensions for every view, add mark comments. 

If possible: standardized our usage of viewWillAppear and viewDidAppear and viewDidLoad.

### PR DESCRIPTION
**Pull request by mindylou**
- Refactors out protocols in view controllers using extensions where possible
- Removes some unnecessary reloading of views in viewWillAppear
- Resolves #419

### CODE REVIEW

File: Podcast/AboutRecastViewController.swift [hunk 1]
```diff
@@ -116,44 +148,18 @@ class AboutRecastViewController: ViewController, UITableViewDataSource, UITableV
             break
```
Comment (AAAstorga) [inline Podcast/AboutRecastViewController.swift:146]: what is case (0,2) that seems like really sketchy. Anyway to make it better?

File: Podcast/AboutRecastViewController.swift [hunk 2]
```diff
@@ -116,44 +148,18 @@ class AboutRecastViewController: ViewController, UITableViewDataSource, UITableV
...
-
         return cell
-
     }
 
-    // MARK: Table View Delegate
+}
+
+// MARK: TableView Delegate
+
```
Comment (AAAstorga) [inline Podcast/AboutRecastViewController.swift]: you don't need a space here

File: Podcast/ActionSheetViewController.swift [hunk 3]
```diff
@@ -171,11 +171,33 @@ class ActionSheetViewController: UIViewController, UITableViewDataSource, UITabl
...
+            
+            self.willMove(toParentViewController: nil)
+            self.view.removeFromSuperview()
+            self.removeFromParentViewController()
+            
+            completion?()
+        })
+    }
+
+    // MARK - TableViewCell Delegate Methods
```
Comment (AAAstorga) [inline Podcast/ActionSheetViewController.swift]: what is this comment referring to?

File: Podcast/ActionSheetViewController.swift [hunk 4]
```diff
@@ -187,43 +209,38 @@ class ActionSheetViewController: UIViewController, UITableViewDataSource, UITabl
         if indexPath.row == options.count - 1 {
```
Comment (AAAstorga) [inline Podcast/ActionSheetViewController.swift:207]: make a comment what you are doing here for other devs

File: Podcast/ActionSheetViewController.swift [hunk 5]
```diff
@@ -246,9 +269,11 @@ class ActionSheetViewController: UIViewController, UITableViewDataSource, UITabl
             dismiss(animated: true)
         }
     }
+
 }
 
-//MARK: - Keyboard
+// MARK: Keyboard
+
```
Comment (AAAstorga) [inline Podcast/ActionSheetViewController.swift]: delete space

File: Podcast/BookmarkViewController.swift [hunk 6]
```diff
@@ -17,9 +15,7 @@ class BookmarkViewController: DiscoverComponentViewController, EmptyStateTableVi
     var continueListeningHeaderView: UIView!
     let continueListeningCollectionViewCellIdentifier: String = "continueListeningIdentifier"
     
-    ///
-    /// Mark: Variables
-    ///
+    // MARK: Variables
```
Comment (AAAstorga) [inline Podcast/BookmarkViewController.swift]: Make more descriptive, variables is v generic

File: Podcast/BookmarkViewController.swift [hunk 7]
```diff
@@ -100,15 +96,70 @@ class BookmarkViewController: DiscoverComponentViewController, EmptyStateTableVi
         super.viewDidAppear(animated)
         fetchContinueListening()
     }
+
+    // MARK: Endpoint Requests
```
Comment (AAAstorga) [inline Podcast/BookmarkViewController.swift]: no space

File: Podcast/BookmarkViewController.swift [hunk 8]
```diff
@@ -100,15 +96,70 @@ class BookmarkViewController: DiscoverComponentViewController, EmptyStateTableVi
...
+            make.width.top.centerX.equalToSuperview()
+            make.height.equalTo(headerViewHeight).priority(999)
+        }
+        bookmarkTableView.layoutIfNeeded()
+    }
+
+}
+
+// MARK: TableView Data Source
+
```
Comment (AAAstorga) [inline Podcast/BookmarkViewController.swift]: space - I know I'm annoying

File: Podcast/BookmarkViewController.swift [hunk 9]
```diff
@@ -156,7 +215,7 @@ class BookmarkViewController: DiscoverComponentViewController, EmptyStateTableVi
         // update index path
```
Comment (AAAstorga) [inline Podcast/BookmarkViewController.swift:211]: delete this comment, seems pointless

File: Podcast/BookmarkViewController.swift [hunk 10]
```diff
@@ -177,62 +236,41 @@ class BookmarkViewController: DiscoverComponentViewController, EmptyStateTableVi
...
-        endpointRequest.failure = { _ in
-            self.bookmarkTableView.endRefreshing()
-            self.bookmarkTableView.stopLoadingAnimation()
-        }
-        System.endpointRequestQueue.addOperation(endpointRequest)
-    }
 }
 
+// MARK: EpisodeDownloader
+
```
Comment (AAAstorga) [inline Podcast/BookmarkViewController.swift]: :P

File: Podcast/BookmarkViewController.swift [hunk 11]
```diff
@@ -177,62 +236,41 @@ class BookmarkViewController: DiscoverComponentViewController, EmptyStateTableVi
...
         // Not worth it, this view doesn't have cells that distinguish download status
-//        if let row = episodes.index(of: episode) {
-//            bookmarkTableView.reloadRows(at: [IndexPath(row: row, section: 0)], with: .none)
-//        }
     }
 }
 
-extension BookmarkViewController: UICollectionViewDataSource, UICollectionViewDelegate, ContinueListeningCollectionViewCellDelegate {
+// MARK: CollectionView Delegate
+
```
Comment (AAAstorga) [inline Podcast/BookmarkViewController.swift]: you get the point ima stop commenting the same thing over and over lol

File: Podcast/BrowseSeriesViewController.swift [hunk 12]
```diff
@@ -135,11 +136,23 @@ class BrowseSeriesViewController: ViewController, UITableViewDataSource, UITable
         return rowHeight
     }
 
+}
+
+// MARK: TableView Delegate
+
+extension BrowseSeriesViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
```
Comment (AAAstorga) [inline Podcast/BrowseSeriesViewController.swift]: thoughts on adding a space here to give it breathing room?

File: Podcast/BrowseSeriesViewController.swift [hunk 13]
```diff
@@ -135,11 +136,23 @@ class BrowseSeriesViewController: ViewController, UITableViewDataSource, UITable
...
+        let seriesDetailViewController = SeriesDetailViewController(series: series[indexPath.row])
+        navigationController?.pushViewController(seriesDetailViewController, animated: true)
+    }
+}
+
+// MARK: SearchSeriesTableView Delegate
+
+extension BrowseSeriesViewController: SearchSeriesTableViewDelegate {
     func searchSeriesTableViewCellDidPressSubscribeButton(cell: SearchSeriesTableViewCell) {
```
Comment (AAAstorga) [inline Podcast/BrowseSeriesViewController.swift]: and here

File: Podcast/BrowseTopicsViewController.swift [hunk 14]
```diff
@@ -9,7 +9,7 @@
 import UIKit
 
 /// Displays a list of topics from either the DiscoverViewController or from a parent BrowseTopicsViewController.
-class BrowseTopicsViewController: ViewController, UITableViewDelegate, UITableViewDataSource {
+class BrowseTopicsViewController: ViewController {
 
     let reuseIdentifier = "Reuse"
```
Comment (AAAstorga) [inline Podcast/BrowseTopicsViewController.swift]: can we make this more descriptive? Reuse is too generic

File: Podcast/BrowseTopicsViewController.swift [hunk 15]
```diff
@@ -80,3 +71,22 @@ class BrowseTopicsViewController: ViewController, UITableViewDelegate, UITableVi
     }
 
 }
+
+// MARK: TableView Delegate
+
+extension BrowseTopicsViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        if let parent = parentTopic { // currently on parent topic, browse the subtopic
```
Comment (AAAstorga) [inline Podcast/BrowseTopicsViewController.swift]: move comment above the line to be consistent

File: Podcast/DiscoverTopicViewController.swift [hunk 16]
```diff
@@ -315,10 +329,10 @@ extension DiscoverTopicViewController: UICollectionViewDelegate, UICollectionVie
...
     }
-
 }
 
-// MARK: - Discover Header View
+// MARK: Discover Header View
+
 extension DiscoverTopicViewController: DiscoverTableViewHeaderDelegate {
     func discoverTableViewHeaderDidPressBrowse(sender: DiscoverCollectionViewHeaderView) {
         switch sender.tag {
```
Comment (AAAstorga) [inline Podcast/DiscoverTopicViewController.swift]: will you eventually have more cases?

File: Podcast/DownloadsViewController.swift [hunk 17]
```diff
@@ -10,16 +10,14 @@ import UIKit
...
+    // MARK: Constants
+    
     var lineHeight: CGFloat = 3
     var topButtonHeight: CGFloat = 30
     var topViewHeight: CGFloat = 60
     
-    ///
-    /// Mark: Variables
-    ///
+    // MARK: Variables
```
Comment (AAAstorga) [inline Podcast/DownloadsViewController.swift:19]: same as earlier

File: Podcast/EpisodeDetailViewController.swift [hunk 18]
```diff
@@ -72,9 +72,13 @@ class EpisodeDetailViewController: ViewController, EpisodeDetailHeaderViewDelega
...
+
+}
+
+// MARK: EpisodeDetailHeaderView Delegate
+
+extension EpisodeDetailViewController: EpisodeDetailHeaderViewDelegate {
+
     func episodeDetailHeaderDidPressRecommendButton(view: EpisodeDetailHeaderView) {
         guard let headerEpisode = episode else { return }
         recast(for: headerEpisode, completion: {_,_ in
```
Comment (AAAstorga) [inline Podcast/EpisodeDetailViewController.swift]: space after `{`

File: Podcast/EpisodeDetailViewController.swift [hunk 19]
```diff
@@ -127,6 +131,8 @@ class EpisodeDetailViewController: ViewController, EpisodeDetailHeaderViewDelega
     
 }
 
+// MARK: Episode Downloader
+
 extension EpisodeDetailViewController: EpisodeDownloader {
     func didReceive(statusUpdate: DownloadStatus, for episode: Episode) {
         if let e = self.episode, e.id == episode.id {
```
Comment (AAAstorga) [inline Podcast/EpisodeDetailViewController.swift]: can we use a better variable for `e`

File: Podcast/FacebookFriendsViewController.swift [hunk 20]
```diff
@@ -155,7 +124,58 @@ class FacebookFriendsViewController: ViewController, UITableViewDelegate, UISear
...
     }
 
-    // MARK: UISearchBarDelegate
+}
+
+// MARK: TableView Data Source
+
+extension FacebookFriendsViewController: UITableViewDataSource {
+
+    func numberOfSections(in tableView: UITableView) -> Int {
```
Comment (AAAstorga) [inline Podcast/FacebookFriendsViewController.swift]: if numberOfSections is 1, you don't need this method. It is default for tableviews

File: Podcast/FeedViewController.swift [hunk 21]
```diff
@@ -10,18 +10,16 @@ import NVActivityIndicatorView
...
-    /// Mark: Constants
-    ///
     var lineHeight: CGFloat = 3
     var topButtonHeight: CGFloat = 30
     var topViewHeight: CGFloat = 60
     
-    ///
-    /// Mark: Variables
-    ///
+    // MARK: Variables
```
Comment (AAAstorga) [inline Podcast/FeedViewController.swift]: same as earlier

File: Podcast/FeedViewController.swift [hunk 22]
```diff
@@ -157,11 +141,54 @@ class FeedViewController: ViewController, FeedElementTableViewCellDelegate  {
...
+        })
+
+        var options = [downloadOption, shareEpisodeOption]
+
+        if case .followingRecommendation = feedElement.context {
+            let bookmarkOption = ActionSheetOption(type: .bookmark(selected: episode.isBookmarked), action: {
+                episode.bookmarkChange()
+            })
+
+            let completion: ((Bool,Int) -> ()) = { isRecommended,_ in
```
Comment (AAAstorga) [inline Podcast/FeedViewController.swift]: should be `isRecommended, _ in`

File: Podcast/FeedViewController.swift [hunk 23]
```diff
@@ -157,11 +141,54 @@ class FeedViewController: ViewController, FeedElementTableViewCellDelegate  {
...
+                self.recast(for: episode, completion: completion)
+            })
+
+            options.insert(bookmarkOption, at: 0)
+            options.insert(recastOption, at: 0)
+        }
+
+
+        if case .followingShare = feedElement.context {
+            if let id = feedElement.id { // only will work when our feedElements contain ids
```
Comment (AAAstorga) [inline Podcast/FeedViewController.swift]: comment above line

File: Podcast/FeedViewController.swift [hunk 24]
```diff
@@ -424,8 +364,71 @@ extension FeedViewController: EmptyStateTableViewDelegate, UITableViewDataSource
...
     }
+
+}
+
+// MARK: TableView Delegate
+
+extension FeedViewController: UITableViewDelegate {
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        if indexPath.section == 0 && showFacebookFriendsSection  && facebookFriends.count > 0 { return }
```
Comment (AAAstorga) [inline Podcast/FeedViewController.swift]: there an extra space after `showFacebookFriendsSection`

File: Podcast/FeedViewController.swift [hunk 25]
```diff
@@ -424,8 +364,71 @@ extension FeedViewController: EmptyStateTableViewDelegate, UITableViewDataSource
...
+    }
+
 }
 
+// MARK: TableView Data Source
+
+extension FeedViewController: UITableViewDataSource {
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return showFacebookFriendsSection && facebookFriends.count > 0 ? 2 : 1
```
Comment (AAAstorga) [inline Podcast/FeedViewController.swift]: use `!facebookFriends.isEmpty`

File: Podcast/FeedViewController.swift [hunk 26]
```diff
@@ -424,8 +364,71 @@ extension FeedViewController: EmptyStateTableViewDelegate, UITableViewDataSource
...
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        if showFacebookFriendsSection && facebookFriends.count > 0 && section == 0 { return 1 }
```
Comment (AAAstorga) [inline Podcast/FeedViewController.swift]: same as above and for any other instances

File: Podcast/FeedViewController.swift [hunk 27]
```diff
@@ -424,8 +364,71 @@ extension FeedViewController: EmptyStateTableViewDelegate, UITableViewDataSource
...
+        switch feedElement.context {
+        case .followingRecommendation(_, let episode), .newlyReleasedEpisode(_, let episode), .followingShare(_, let episode):
+            if episode.isPlaying {
+                currentlyPlayingIndexPath = indexPath
+            }
+        case .followingSubscription:
+            break
+        }
+        var cell = feedTableView.dequeueReusableCell(withIdentifier: feedElement.context.cellType.identifier) as! UITableViewCell & FeedElementTableViewCell
+        if let recastCell = cell as? FeedRecastTableViewCell { // adjust cells expansion
```
Comment (AAAstorga) [inline Podcast/FeedViewController.swift]: comment goes above

File: Podcast/FollowerFollowingViewController.swift [hunk 28]
```diff
@@ -84,10 +75,12 @@ class FollowerFollowingViewController: ViewController, UITableViewDataSource, UI
...
-    // MARK: TableView Delegate & DataSource Methods
-    //
+
+}
+
+// MARK: TableView Data Source
+
+extension FollowerFollowingViewController: UITableViewDataSource {
 
     func numberOfSections(in tableView: UITableView) -> Int {
```
Comment (AAAstorga) [inline Podcast/FollowerFollowingViewController.swift]: delete function

File: Podcast/FollowerFollowingViewController.swift [hunk 29]
```diff
@@ -115,15 +101,43 @@ class FollowerFollowingViewController: ViewController, UITableViewDataSource, UI
...
+
+// MARK: EmptyStateTableView Delegate
+
+extension FollowerFollowingViewController: EmptyStateTableViewDelegate {
+
+    func emptyStateTableViewHandleRefresh() {
+        fetchUsers()
+    }
+
+    func didPressEmptyStateViewActionItem() {
```
Comment (AAAstorga) [inline Podcast/FollowerFollowingViewController.swift]: make this all one line if nothing in it and remove comment

File: Podcast/InternalProfileViewController.swift [hunk 30]
```diff
@@ -133,9 +133,34 @@ class InternalProfileViewController: ViewController, UITableViewDelegate, UITabl
...
+
+    func fetchSubscriptions() {
+
+        guard let userID = System.currentUser?.id else { return }
+
+        let userSubscriptionEndpointRequest = FetchUserSubscriptionsEndpointRequest(userID: userID)
+
+        userSubscriptionEndpointRequest.success = { (endpointRequest: EndpointRequest) in
+            guard let subscriptions = endpointRequest.processedResponseValue as? [Series] else { return }
+            self.subscriptions = subscriptions.sorted { $0.lastUpdated ?? NSDate.distantPast  > $1.lastUpdated ?? NSDate.distantPast}
```
Comment (AAAstorga) [inline Podcast/InternalProfileViewController.swift]: extra space after `NSDate.distantPast` and can we use `Date.distantPast`?

File: Podcast/InternalProfileViewController.swift [hunk 31]
```diff
@@ -133,9 +133,34 @@ class InternalProfileViewController: ViewController, UITableViewDelegate, UITabl
...
+            self.subscriptionsTableView.reloadData()
+            self.remakeSubscriptionTableViewContraints()
+        }
+
+        userSubscriptionEndpointRequest.failure = { (endpointRequest: EndpointRequest) in
+        }
```
Comment (AAAstorga) [inline Podcast/InternalProfileViewController.swift]: change to `_ in` if you don't use endpointRequest

File: Podcast/InternalProfileViewController.swift [hunk 32]
```diff
@@ -145,9 +170,54 @@ class InternalProfileViewController: ViewController, UITableViewDelegate, UITabl
...
+                navigationController?.pushViewController(FacebookFriendsViewController(), animated: true)
+            case .bookmark:
+                tabBarController.selectedIndex = System.bookmarkTab
+                break
+            case .shared:
+                navigationController?.pushViewController(SharedContentViewController(), animated: true)
+            }
+        case subscriptionsTableView:
+            guard let subs = subscriptions else { return }
+            if subs.count > 0 {
```
Comment (AAAstorga) [inline Podcast/InternalProfileViewController.swift]: use `!subs.isEmpty`

File: Podcast/ListeningHistoryViewController.swift [hunk 33]
```diff
@@ -8,19 +8,17 @@
...
     var topButtonHeight: CGFloat = 30
     var topViewHeight: CGFloat = 60
     
-    ///
-    /// Mark: Variables
-    ///
-    var listeningHistoryTableView: EmptyStateTableView! //not a delegate because no action button
+    // MARK: Variables
+    
+    var listeningHistoryTableView: EmptyStateTableView! // not a delegate because no action button
```
Comment (AAAstorga) [inline Podcast/ListeningHistoryViewController.swift]: comment goes above

File: Podcast/ListeningHistoryViewController.swift [hunk 34]
```diff
@@ -58,11 +56,64 @@ class ListeningHistoryViewController: ViewController, UITableViewDelegate, UITab
...
-    //MARK: -
-    //MARK: TableView DataSource
-    //MARK: -
-    
+    func emptyStateTableViewHandleRefresh() {
+        fetchEpisodes()
+    }
+
+    func fetchEpisodes(refresh: Bool = true) {
+        let offset: Int
```
Comment (AAAstorga) [inline Podcast/ListeningHistoryViewController.swift]: `let offset = refresh ? 0 : self.episodes.count`

File: Podcast/ListeningHistoryViewController.swift [hunk 35]
```diff
@@ -58,11 +56,64 @@ class ListeningHistoryViewController: ViewController, UITableViewDelegate, UITab
...
+            offset = 0
+        } else {
+            offset = self.episodes.count
+        }
+        let historyRequest = FetchListeningHistoryEndpointRequest(offset: offset, max: pageSize)
+        historyRequest.success = { request in
+            guard let newEpisodes = request.processedResponseValue as? [Episode] else { return }
+            var episodesToAdd: [Episode] = []
+            for episode in newEpisodes {
+                if !self.episodeSet.contains(episode) { //only add episodes we haven't seen
```
Comment (AAAstorga) [inline Podcast/ListeningHistoryViewController.swift]: comments goes above here and line 83

File: Podcast/PlayerViewController.swift [hunk 36]
```diff
@@ -284,9 +228,99 @@ class PlayerViewController: TabBarAccessoryViewController, PlayerDelegate, Playe
...
+
+    func playerHeaderViewDidDrag(sender: UIPanGestureRecognizer) {
+        let touchPoint = sender.location(in: view.window)
+        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else { return }
+
+        switch sender.state {
+        case .began:
+            initialTouchPoint = touchPoint
+        case .changed:
+            if touchPoint.y > 0 && touchPoint.y < view.frame.height - appDelegate.tabBarController.tabBar.frame.height - miniPlayerView.miniPlayerHeight {
```
Comment (AAAstorga) [inline Podcast/PlayerViewController.swift]: def put a nice comment here and what is going on

File: Podcast/SearchDiscoverViewController.swift [hunk 37]
```diff
@@ -221,7 +230,77 @@ class SearchDiscoverViewController: ViewController, UISearchControllerDelegate,
         pastSearchesTableView.reloadData()
     }
 
-    // MARK: - Search Delegate
+    func addPastSearches() {
+        guard let searchText = searchController.searchBar.text else { return }
+        if searchText == "" { return }
```
Comment (AAAstorga) [inline Podcast/SearchDiscoverViewController.swift]: `searchText.isEmpty`

File: Podcast/SeriesDetailViewController.swift [hunk 38]
```diff
@@ -154,35 +154,22 @@ class SeriesDetailViewController: ViewController, SeriesDetailHeaderViewDelegate
...
-        let activityViewController = UIActivityViewController(activityItems: [], applicationActivities: nil)
-        present(activityViewController, animated: true, completion: nil)
+    func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        let scale: CGFloat = 50.0
+        let offset = max(0, -(scrollView.contentOffset.y + scrollView.adjustedContentInset.top))
+        let scaledOffset = offset / scale
+
+        seriesHeaderView.infoView.alpha = 1.0 - scaledOffset
+        seriesHeaderView.contentContainerTop?.update(offset: -offset)
+        seriesHeaderView.gradientView.alpha = 1.85 - scaledOffset * 0.75
```
Comment (AAAstorga) [inline Podcast/SeriesDetailViewController.swift:164]: why these numbers. Comment explaining why you use such specifics

File: Podcast/SeriesDetailViewController.swift [hunk 39]
```diff
@@ -207,25 +194,56 @@ class SeriesDetailViewController: ViewController, SeriesDetailHeaderViewDelegate
...
+    //create and delete subscriptions
+    func seriesDetailHeaderViewDidPressSubscribeButton(seriesDetailHeader: SeriesDetailHeaderView) {
+        series!.subscriptionChange(completion: seriesDetailHeader.subscribeButtonChangeState)
+    }
+
+    func seriesDetailHeaderViewDidPressMoreTopicsButton(seriesDetailHeader: SeriesDetailHeaderView) {
+        // Show view of all tags?
+    }
+
+    func seriesDetailHeaderViewDidPressSettingsButton(seriesDetailHeader: SeriesDetailHeaderView) {
```
Comment (AAAstorga) [inline Podcast/SeriesDetailViewController.swift]: can compact this function into one line

File: Podcast/SubscriptionsViewController.swift [hunk 40]
```diff
@@ -103,9 +72,42 @@ class SubscriptionsViewController: ViewController, UICollectionViewDelegate, UIC
...
-    //MARK:
-    //MARK: - Empty state view delegate
-    //MARK:
+}
+
+// MARK: CollectionView Data Source
+
+extension SubscriptionsViewController: UICollectionViewDataSource {
+
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
```
Comment (AAAstorga) [inline Podcast/SubscriptionsViewController.swift]: delete this function

File: Podcast/UserDetailViewController.swift [hunk 41]
```diff
@@ -254,7 +253,12 @@ extension UserDetailViewController: UITableViewDelegate, UITableViewDataSource {
...
     }
-    
+
+}
+
+// MARK: TableView Delegate
+
+extension UserDetailViewController: UITableViewDelegate {
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
         if recasts.count > 0 {
```
Comment (AAAstorga) [inline Podcast/UserDetailViewController.swift]: use isEmpty

File: Podcast/UserDetailViewController.swift [hunk 42]
```diff
@@ -266,18 +270,16 @@ extension UserDetailViewController: UITableViewDelegate, UITableViewDataSource {
...
-//
-extension UserDetailViewController: UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
-    
+
+// MARK: CollectionView Data Source
+
+extension UserDetailViewController: UICollectionViewDataSource {
+
     func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
         return subscriptions.count == 0 ? 1 : subscriptions.count
```
Comment (AAAstorga) [inline Podcast/UserDetailViewController.swift]: isEmpty

File: Podcast/UserDetailViewController.swift [hunk 43]
```diff
@@ -290,7 +292,12 @@ extension UserDetailViewController: UICollectionViewDataSource, UICollectionView
...
     }
-    
+
+}
+
+// MARK: CollectionView Delegate Flow Layout
+
+extension UserDetailViewController: UICollectionViewDelegateFlowLayout {
     func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
         if subscriptions.count > 0 {
```
Comment (AAAstorga) [inline Podcast/UserDetailViewController.swift]: isEmpty

File: Podcast/UserDetailViewController.swift [hunk 44]
```diff
@@ -266,18 +270,16 @@ extension UserDetailViewController: UITableViewDelegate, UITableViewDataSource {
...
+
+extension UserDetailViewController: UICollectionViewDataSource {
+
     func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
         return subscriptions.count == 0 ? 1 : subscriptions.count
     }
-    
+
     func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
         if subscriptions.count > 0 {
```
Comment (AAAstorga) [inline Podcast/UserDetailViewController.swift]: isEmpty

File: Podcast/UserDetailViewController.swift [hunk 45]
```diff
@@ -300,7 +307,11 @@ extension UserDetailViewController: UICollectionViewDataSource, UICollectionView
...
         }
     }
-    
+}
+
+// MARK: CollectionView Delegate
+
+extension UserDetailViewController: UICollectionViewDelegate {
     func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
         if subscriptions.count > 0 {
```
Comment (AAAstorga) [inline Podcast/UserDetailViewController.swift]: you betcha, isEmpty

**Review by AAAstorga — COMMENTED** [review]
Lots of comments, but all minor things. Mostly styling and commenting some areas that look confusing. Big PR but awesome work!!!

File: Podcast/SeriesDetailViewController.swift (hunk 38 above)
Comment (AAAstorga) [reply Podcast/SeriesDetailViewController.swift:164]: @dantheli

File: Podcast/FollowerFollowingViewController.swift (hunk 29 above)
Comment (mindylou) [reply Podcast/FollowerFollowingViewController.swift]: it's a required method in the protocol so i think that's why the comment is there to specify why it's a noop

File: Podcast/PlayerViewController.swift [hunk 46]
```diff
@@ -284,9 +226,98 @@ class PlayerViewController: TabBarAccessoryViewController, PlayerDelegate, Playe
...
+
+        switch sender.state {
+        case .began:
+            initialTouchPoint = touchPoint
+        case .changed:
+            // if the point touched on the player is within the top of the mini player and the top of the screen,
+            // adjust the player's height/expansion and transparency
+            if touchPoint.y > 0 && touchPoint.y < view.frame.height - appDelegate.tabBarController.tabBar.frame.height - miniPlayerView.miniPlayerHeight {
+                view.frame = CGRect(x: 0, y: touchPoint.y, width: view.frame.width, height: view.frame.height)
+                episodeDetailView.alpha = 1 - (touchPoint.y/view.frame.height)
```
Comment (doctor3w) [inline Podcast/PlayerViewController.swift]: Can you add spaces between operators here? And same below?

File: Podcast/PlayerViewController.swift [hunk 47]
```diff
@@ -284,9 +226,98 @@ class PlayerViewController: TabBarAccessoryViewController, PlayerDelegate, Playe
...
+    func miniPlayerViewDidDrag(sender: UIPanGestureRecognizer) {
+        let touchPoint = sender.location(in: view.window)
+        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else { return }
+
+        switch sender.state {
+        case .began:
+            initialTouchPoint = touchPoint
+        case .changed:
+            if touchPoint.y < view.frame.height - appDelegate.tabBarController.tabBar.frame.height - miniPlayerView.miniPlayerHeight {
+                episodeDetailView.alpha = 1 - (touchPoint.y/view.frame.height)
```
Comment (doctor3w) [inline Podcast/PlayerViewController.swift:295]: SAme here

**Review by doctor3w — CHANGES_REQUESTED** [review]

**Review by doctor3w — APPROVED** [review]

**Review by AAAstorga — APPROVED** [review]

**Pull request merged**